### PR TITLE
Separate RadialBasis into a separate static library

### PR DESCRIPTION
**Pull request by mortenpi**
This (rather large) PR is the minimum that I was able to put together to separate out `RadialBasis`, its dependencies, and a few other functions into a separate CMake target that compiles into a static library.

It's then used in [this script](https://github.com/mortenpi/libhelfem_jll.jl/blob/source/build_tarballs.jl) to build a shared library ([CxxWrap wrapper](https://github.com/mortenpi/libhelfem_jll.jl/blob/source/src/libhelfem-cxxwrap.cc)) that powers the [HelFEM.jl](https://github.com/mortenpi/HelFEM.jl/tree/mp/jll) Julia wrapper package.

Essentially what I did though:

* Move the minimum number of libraries (headers and implementation) into `libhelfem/`.
* The classes and functions that I use in HelFEM.jl live as public headers in `libhelfem/include`.
* Headers that had multiple classes in them I split up into separate files, to make the dependency graph clearer.
* I also created a single header file that includes all the libraries so that you can just do `include <helfem.h>` to get all the public `helfem` functions.
* In the application programs under `src/` I made minimal changes -- basically, just changing some `include`s here and there to keep everything working.
* The CMake scripts now use the `find_package` to find the Armadillo and GSL libraries.
* I changed the `CMake.system` logic slightly as well -- it is now optional and only gets included if the file exists. By default, there is no `CMake.system` in the repository, and it is a `.gitignore`d file.
* I am also using `clang-format` to auto-format the source to have a consistent style (but now I think it's only the `helfem.h`).

TODO:

- [x] Do not use `"` for directories in CMake files
- [x] `sap_effective_charge` don't merge with machine generated file
- [ ] `FindArmadillo` header only
- [x] `Sinvh` not returning warning?
- [x] Switch to `NamespaceIndentation: All` for namespace indentation in clang-format.

### CODE REVIEW

File: src/general/scf_helpers.h [hunk 1]
```diff
@@ -47,7 +48,9 @@ namespace helfem {
     /// Form natural orbitals
     void form_NOs(const arma::mat & P, const arma::mat & Sh, const arma::mat & Sinvh, arma::mat & AO_to_NO, arma::mat & NO_to_AO, arma::vec & occs);
     /// Form half-inverse overlap
-    arma::mat form_Sinvh(arma::mat S, bool chol=false);
+    inline arma::mat form_Sinvh(arma::mat S, bool chol=false) {
+      utils::invh(S, chol);
```
Comment (mortenpi) [inline src/general/scf_helpers.h:52]:
```suggestion
      return utils::invh(S, chol);
```

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,2 @@
+BasedOnStyle: LLVM
+ColumnLimit: 92
```
Comment (mortenpi) [inline .clang-format:2]:
```suggestion
ColumnLimit: 80
```

File: libhelfem/src/helfem.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ *                             -
+ * Finite element methods for electronic structure calculations on small
+ * systems
+ *
+ * Copyright (c) 2018 Susi Lehtola
+ *
+ * This program is free software; you can redistribute it and/or * modify it
+ * under the terms of the GNU General Public License * as published by the Free
+ * Software Foundation; either version 2 * of the License, or (at your option)
+ * any later version.
```
Comment (mortenpi) [inline libhelfem/src/helfem.cpp:14]: Asterisks in wrong places.

File: CMakeLists.txt [hunk 4]
```diff
@@ -2,6 +2,34 @@
...
+#     cmake .. -DCMAKE_PREFIX_PATH=$HOME/local/usr
+#
+find_package(Armadillo REQUIRED)
+message("Found Armadillo ${ARMADILLO_VERSION_NAME} (v${ARMADILLO_VERSION_MAJOR}.${ARMADILLO_VERSION_MINOR}.${ARMADILLO_VERSION_PATCH})")
+message("ARMADILLO_VERSION_STRING: ${ARMADILLO_VERSION_STRING}")
+message("ARMADILLO_INCLUDE_DIRS:   ${ARMADILLO_INCLUDE_DIRS}")
+message("ARMADILLO_LIBRARIES:      ${ARMADILLO_LIBRARIES}")
+if(NOT (${ARMADILLO_VERSION_MAJOR} EQUAL "9"))
+    message(FATAL_ERROR "Invalid Armadillo version, expected 9.")
+endif()
```
Comment (mortenpi) [inline CMakeLists.txt:22]: Remove this version check.

File: CMakeLists.txt [hunk 5]
```diff
@@ -13,7 +41,11 @@ if(USE_OPENMP)
 endif()
 
 # System specific definitions
-include(CMake.system)
+if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/CMake.system")
+    message("Including user-provided CMake.system")
+    include("${CMAKE_CURRENT_SOURCE_DIR}/CMake.system")
+endif()
```
Comment (susilehtola) [inline CMakeLists.txt:47]: This check should be done above. If CMake.system exists, then use the definitions therein; only try searching for Armadillo and GSL if CMake.system does not exist.

**Comment by mortenpi** [comment]
Implemented what we discussed in the morning, but a few things need another round of review:

1. To make sure I am not modifying the generated `sap.cpp/h` files, I decided to remove the vectorized version of `sap_effective_charge` altogether (3d65001). The vectorized version is not actually used anywhere (anymore?) and would be easy to add back if it's ever needed again?

2. I more or less re-engineered the CMake files (ed97a41). This definitely needs testing to make sure it works with your workflow as well.

   1. The default behavior is now similar to what it is on master -- CMake does not set up any libraries for you (other than OpenMP). A `CMake.system` file is necessary to make the build succeed.

   2. I removed the `target_link_libraries` calls. The current expectation is that the user should set up all the necessary libraries with (global) `link_libraries` calls in `CMake.system`. I made a corresponding change in `CMake.dx7-lehtola`, but I don't actually know if it works.

      On my system, I can (without using any of the `find_package` calls) configure and build HeFEM successfully with the following `CMake.system` file:
 
      ```cmake
      # With Armadillo and libxc installed under deps/usr:
      include_directories("deps/usr/include" "/usr/include/hdf5/serial/")
      link_directories("${CMAKE_CURRENT_SOURCE_DIR}/deps/usr/lib")
      link_libraries(gsl blas lapack armadillo hdf5_serial xc)
      ```

      I _think_ the reason I need to link against the Armadillo shared library is because I didn't disable the shared library in my build, so I need the Armadillo wrapper. I assume if the Armadillo build has the shared library disabled, it could be omitted from `link_libraries`.

   3. The get the behavior need/want for the Julia wrapper, I can just pass some options to `cmake` (specifically `-DHELFEM_BINARIES=OFF -DHELFEM_FIND_DEPS=ON`). The new options [are  documented in `CMakeLists.txt`](https://github.com/mortenpi/HelFEM/blob/ed97a4108215a350e784b55a5a1b6fd1b42512d7/CMakeLists.txt#L5-L30).

   4. I added a very trivial CMake module that tries to automatically find libxc (only relevant if `-DHELFEM_FIND_DEPS=ON` is passed).

3. I kept the Armadillo version check, but with the caveat that it's only done when `find_package` is used (i.e. `-DHELFEM_FIND_DEPS=ON`). Also, I changed it from `== 9` to `>= 9`, so it would be compatible with future Armadillo versions.

**Comment by susilehtola** [comment]
Forgot about this again. I think the changes were okay; I can anyway make any changes later on if they turn out to be necessary.

Thanks

**Pull request merged**